### PR TITLE
add:impl builtin uncompress & uncompressedlength

### PR DESCRIPTION
**Pull request by woodpenker**
1. impl builtin func uncompress & uncompressed_length func.
2. fix a small mistake of compress func to get "Empty strings are stored
as empty strings."
3. add more test cases.

### CODE REVIEW

File: plan/typeinferer_test.go [hunk 1]
```diff
@@ -265,6 +265,8 @@ func (ts *testTypeInferrerSuite) TestInferType(c *C) {
 		{`aes_decrypt("pingcap", "fit2cloud@2014")`, mysql.TypeVarString, charset.CharsetUTF8, 0},
 		{`md5(123)`, mysql.TypeVarString, charset.CharsetUTF8, 0},
 		{`compress('love')`, mysql.TypeBlob, charset.CharsetBin, mysql.BinaryFlag},
+		{`uncompress('love')`, mysql.TypeVarString, charset.CharsetUTF8, 0},
```
Comment (shenli) [inline plan/typeinferer_test.go]: Uncompress returns LONGBLOB with charset binary.

File: expression/builtin_encryption.go [hunk 2]
```diff
@@ -212,7 +213,11 @@ func (b *builtinCompressSig) eval(row []types.Datum) (d types.Datum, err error)
 	if err != nil {
 		return d, errors.Trace(err)
 	}
-
+	//Empty strings are stored as empty strings.
```
Comment (shenli) [inline expression/builtin_encryption.go]: Leave a blank after //

File: expression/builtin_encryption.go [hunk 3]
```diff
@@ -520,7 +525,35 @@ type builtinUncompressSig struct {
 
 // See https://dev.mysql.com/doc/refman/5.7/en/encryption-functions.html#function_uncompress
 func (b *builtinUncompressSig) eval(row []types.Datum) (d types.Datum, err error) {
-	return d, errFunctionNotExists.GenByArgs("UNCOMPRESS")
+
```
Comment (XuHuaiyu) [inline expression/builtin_encryption.go]: remove this line

File: expression/builtin_encryption.go [hunk 4]
```diff
@@ -520,7 +525,35 @@ type builtinUncompressSig struct {
...
+		return d, errors.Trace(err)
+	}
+	if len(compressStr) == 0 {
+		d.SetString("")
+		return d, nil
+	}
+	reader := bytes.NewReader(compressStr)
+	var out bytes.Buffer
+	r, err := zlib.NewReader(reader)
+
```
Comment (XuHuaiyu) [inline expression/builtin_encryption.go]: remove this line

File: expression/builtin_encryption.go [hunk 5]
```diff
@@ -537,7 +570,36 @@ type builtinUncompressedLengthSig struct {
 
 // See https://dev.mysql.com/doc/refman/5.7/en/encryption-functions.html#function_uncompressed-length
 func (b *builtinUncompressedLengthSig) eval(row []types.Datum) (d types.Datum, err error) {
-	return d, errFunctionNotExists.GenByArgs("UNCOMPRESSED_LENGTH")
+
```
Comment (XuHuaiyu) [inline expression/builtin_encryption.go]: ditto.

File: expression/builtin_encryption.go [hunk 6]
```diff
@@ -537,7 +570,36 @@ type builtinUncompressedLengthSig struct {
...
+		return d, errors.Trace(err)
+	}
+	if len(compressStr) == 0 {
+		d.SetInt64(0)
+		return d, nil
+	}
+	reader := bytes.NewReader(compressStr)
+	var out bytes.Buffer
+	r, err := zlib.NewReader(reader)
+
```
Comment (XuHuaiyu) [inline expression/builtin_encryption.go]: ditto

**Comment by XuHuaiyu** [comment]
@woodpenker Please address the comments~

File: expression/builtin_encryption.go (hunk 4 above)
Comment (woodpenker) [inline expression/builtin_encryption.go]: ridiculous

**Comment by woodpenker** [comment]
@XuHuaiyu Already  extract the duplicated code into a function

File: expression/builtin_encryption.go [hunk 7]
```diff
@@ -523,14 +523,24 @@ type builtinUncompressSig struct {
 	baseBuiltinFunc
 }
 
+func uncompress(compressStr []byte) ([]byte, error) {
+	reader := bytes.NewReader(compressStr)
+	var out bytes.Buffer
+	r, err := zlib.NewReader(reader)
+	if err != nil {
+		return nil, nil
```
Comment (XuHuaiyu) [inline expression/builtin_encryption.go]: Should we `return nil, errors.Trace(err)`?

File: expression/builtin_encryption.go [hunk 8]
```diff
@@ -523,14 +523,24 @@ type builtinUncompressSig struct {
 	baseBuiltinFunc
 }
 
+func uncompress(compressStr []byte) ([]byte, error) {
```
Comment (XuHuaiyu) [inline expression/builtin_encryption.go]: comment for this func.

File: expression/builtin_encryption.go [hunk 9]
```diff
@@ -543,16 +553,11 @@ func (b *builtinUncompressSig) eval(row []types.Datum) (d types.Datum, err error
...
 	}
-	reader := bytes.NewReader(compressStr)
-	var out bytes.Buffer
-	r, err := zlib.NewReader(reader)
-
-	if err != nil {
-		return d, nil
+	bytes, err := uncompress(compressStr)
+	if err != nil || bytes == nil {
+		return d, err
```
Comment (XuHuaiyu) [inline expression/builtin_encryption.go]: s/err/errors.Trace(err)

File: expression/builtin_encryption.go [hunk 10]
```diff
@@ -570,35 +575,27 @@ type builtinUncompressedLengthSig struct {
...
 	}
-	reader := bytes.NewReader(compressStr)
-	var out bytes.Buffer
-	r, err := zlib.NewReader(reader)
-
-	if err != nil {
-		return d, errors.Trace(err)
+	bytes, err := uncompress(compressStr)
+	if err != nil || bytes == nil {
+		return d, err
```
Comment (XuHuaiyu) [inline expression/builtin_encryption.go]: return d, errors.Trace(err)

File: expression/builtin_encryption.go [hunk 11]
```diff
@@ -570,35 +575,27 @@ type builtinUncompressedLengthSig struct {
...
 		return d, nil
 	}
-
 	compressStr, err := arg.ToBytes()
 	if err != nil {
 		return d, errors.Trace(err)
 	}
 	if len(compressStr) == 0 {
-		d.SetInt64(0)
+		d.SetString("")
```
Comment (XuHuaiyu) [inline expression/builtin_encryption.go]: why set d to string?

**Comment by XuHuaiyu** [comment]
@woodpenker hi, any updates? 
Please resolve the conflicts.

**Comment by woodpenker** [comment]
@XuHuaiyu updated now

File: expression/builtin_encryption.go [hunk 12]
```diff
@@ -523,12 +523,13 @@ type builtinUncompressSig struct {
 	baseBuiltinFunc
 }
 
+// This func is used for umcopress a string compiled with a compression library such as zlib.
```
Comment (XuHuaiyu) [inline expression/builtin_encryption.go]: s/uncopress/uncompress

**Comment by XuHuaiyu** [comment]
LGTM 
PTAL @shenli @zimulala @tiancaiamao @hanfei1991

File: expression/builtin_encryption.go [hunk 13]
```diff
@@ -553,7 +592,28 @@ type builtinUncompressedLengthSig struct {
...
+	compressStr, err := arg.ToBytes()
+	if err != nil {
+		return d, errors.Trace(err)
+	}
+	if len(compressStr) == 0 {
+		d.SetInt64(0)
+		return d, nil
+	}
+	bytes, err := uncompress(compressStr)
+	if err != nil || bytes == nil {
```
Comment (hanfei1991) [inline expression/builtin_encryption.go]: ditto

File: expression/builtin_encryption.go [hunk 14]
```diff
@@ -532,10 +537,44 @@ type builtinUncompressSig struct {
...
+	compressStr, err := arg.ToBytes()
+	if err != nil {
+		return d, errors.Trace(err)
+	}
+	if len(compressStr) == 0 {
+		d.SetString("")
+		return d, nil
+	}
+	bytes, err := uncompress(compressStr)
+	if err != nil || bytes == nil {
```
Comment (hanfei1991) [inline expression/builtin_encryption.go]: When the bytes is nil ? Do we really check this?

Comment (woodpenker) [inline expression/builtin_encryption.go]: YEAH,it's no need...

**Pull request merged**